chainerrl/agents/iqn.py

```diff
@@ -0,0 +1,355 @@
+            n_actions = h.shape[-1]
+            h = F.reshape(h, (batch_size, n_taus, n_actions))
+            return QuantileDiscreteActionValue(h)
+
+        return evaluate_with_quantile_thresholds
+
+
+def _unwrap_variable(x):
+    if isinstance(x, chainer.Variable):
+        return x.array
+    else:
+        return x
+
+
+def compute_eltwise_huber_quantile_loss(y, t, taus, huber_loss_threshold=1.0):
+    """Compute elementwise Huber losses for quantile regression.
+
+    Args:
+        y (chainer.Variable): (batch_size, N)
+        t (chainer.Variable or ndarray): (batch_size, N_prime)
+        taus (ndarray): (batch_size, N)
+        huber_loss_threshold (float): Threshold of Huber loss. In the IQN
+            paper, this is denoted by kappa.
+
+    Returns:
+        chainer.Variable: Loss (batch_size, N, N_prime)
+    """
+    assert y.shape == taus.shape
+    # (batch_size, N) -> (batch_size, N, 1)
+    y = F.expand_dims(y, axis=2)
+    # (batch_size, N_prime) -> (batch_size, 1, N_prime)
+    t = F.expand_dims(t, axis=1)
+    # (batch_size, N) -> (batch_size, N, 1)
+    taus = F.expand_dims(taus, axis=2)
+    # Broadcast to (batch_size, N, N_prime)
+    y, t, taus = F.broadcast(y, t, taus)
+    with chainer.no_backprop_mode():
+        I_delta = ((_unwrap_variable(t) - _unwrap_variable(y)) > 0).astype('f')
+    eltwise_huber_loss = F.huber_loss(
+        y, t, delta=huber_loss_threshold, reduce='no')
+    eltwise_loss = abs(taus - I_delta) * eltwise_huber_loss
+    return eltwise_loss
+
+
+class IQN(dqn.DQN):
+    """Implicit Quantile Networks.
+
+    See https://arxiv.org/abs/1806.06923.
+
+    Args:
+        quantile_thresholds_N (int): Number of quantile thresholds used in
+            quantile regression.
+        quantile_thresholds_N_prime (int): Number of quantile thresholds used
+            to sample from the return distribution at the next state.
+        quantile_thresholds_K (int): Number of quantile thresholds used to
+            compute greedy actions.
+
+    For other arguments, see chainerrl.agents.DQN.
+    """
+
+    def __init__(self, *args, **kwargs):
+        # N=N'=64 and K=32 were used in the IQN paper's experiments
+        # (personal communication)
+        self.quantile_thresholds_N = kwargs.pop('quantile_thresholds_N', 64)
+        self.quantile_thresholds_N_prime = kwargs.pop(
+            'quantile_thresholds_N_prime', 64)
+        self.quantile_thresholds_K = kwargs.pop('quantile_thresholds_K', 32)
+        super().__init__(*args, **kwargs)
+
+    def _compute_target_values(self, exp_batch):
+        """Compute a batch of target return distributions.
+
+        Returns:
+            chainer.Variable: (batch_size, N_prime).
+        """
+        batch_next_state = exp_batch['next_state']
+        batch_size = batch_next_state.shape[0]
+        taus_tilde = self.xp.random.uniform(
+            0, 1, size=(batch_size, self.quantile_thresholds_K)).astype('f')
```

Comment:
`cupy.random.*` creates an array on the current device, which sometimes causes an error.  Have you tested `--gpu 1`?

Reply:
The current device is set by `DQN.__init__`. I confirmed that `python examples/ale/train_iqn_ale.py --gpu 1` worked.